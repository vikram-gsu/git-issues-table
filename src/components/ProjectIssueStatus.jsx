import React, { Component } from "react";
import { Table, Grid, Search } from "semantic-ui-react";
import "./styles/ProjectIssueStatus.css";


const Labels = ({ labels }) => (
  <ul>
    {labels.map(({ url, color, name, id }) => (
      <li key={id}>
        <a
          href={url}
          style={{
            backgroundColor: `#${color}`,
            color: "black",
            padding: "0.1em",
            borderRadius: "1px"
          }}
        >
          {name}
        </a>
      </li>
    ))}
  </ul>
);

const escape = s => {
  return s.replace(/[-\\^$*+?.()|[\]{}]/g, "\\$&");
};

class ProjectIssueStatus extends Component {
  initialState = {
    issues: this.props.issues,
    isLoading: false,
    value: ""
  };
  state = this.initialState;

  handleSearchChange = (e, { value }) => {
    this.setState({ isLoading: true, value });

    // console.log(this.state.issues.filter(issue => isMatch(issue)))
    setTimeout(() => {
      if (this.state.value.length < 1) return this.setState(this.initialState);

      const re = new RegExp(escape(this.state.value), "i");
      const isMatch = result => re.test(result.title);

      this.setState({
        isLoading: false,
        issues: this.props.issues.filter(issue => isMatch(issue))
      });
    }, 300);
  };

  render() {
    return (
      <div>
        <Grid columns={16}>
          <Grid.Row>
            <Grid.Column width={2}></Grid.Column>
            <Grid.Column width={4}>
              <h1>Github Issue Details</h1>
            </Grid.Column>
            <Grid.Column width={4}></Grid.Column>
            <Grid.Column width={4}>
              <div className="search">
                <Search
                  size="huge"
                  width="100%"
                  loading={this.state.isLoading}
                  onSearchChange={this.handleSearchChange}
                />
              </div>
            </Grid.Column>
            <Grid.Column width={2}></Grid.Column>
          </Grid.Row>
          <Grid.Column width={2}></Grid.Column>
          <Grid.Column width={12}>
            <div className="table">
              <Table celled>
                <Table.Header>
                  <Table.Row>
                    <Table.HeaderCell>Issue Number</Table.HeaderCell>
                    <Table.HeaderCell>Title</Table.HeaderCell>
                    <Table.HeaderCell>Created At</Table.HeaderCell>
                    <Table.HeaderCell>Updated At</Table.HeaderCell>
                    <Table.HeaderCell>Labels</Table.HeaderCell>
                    <Table.HeaderCell>State</Table.HeaderCell>
                  </Table.Row>
                </Table.Header>
                <Table.Body>
                  {this.state.issues &&
                    this.state.issues.map(issue => {
                      const {
                        number: IssueNumber,
                        title,
                        created_at,
                        updated_at,
                        labels,
                        state
                      } = issue;
                      return (
                        <Table.Row key={IssueNumber}>
                          <Table.Cell>{IssueNumber}</Table.Cell>
                          <Table.Cell>{title}</Table.Cell>
                          <Table.Cell>{created_at}</Table.Cell>
                          <Table.Cell>{updated_at}</Table.Cell>
                          <Table.Cell>
                            <Labels labels={labels} />
                          </Table.Cell>
                          <Table.Cell>{state}</Table.Cell>
                        </Table.Row>
                      );
                    })}
                </Table.Body>
              </Table>
            </div>
          </Grid.Column>
          <Grid.Column width={2}></Grid.Column>
        </Grid>
      </div>
    );
  }
}

export default ProjectIssueStatus;
