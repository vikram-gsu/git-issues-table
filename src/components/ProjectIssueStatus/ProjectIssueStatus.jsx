import React, { Component } from "react";
import { Table } from "semantic-ui-react";

const captured_response = require("../../data/captured_response.json");

const Labels = ({ labels }) => (
  <ul>
    {labels.map(({ url, color, name, id }) => (
      <li key={id}>
        <a href={url} style={{ color: `#${color}` }}>
          {name}
        </a>
      </li>
    ))}
  </ul>
);

const formatDate = dateString => {
  const dateObj = new Date(dateString);
  const options = {
    year: "numeric",
    month: "long",
    day: "numeric",
    hour: "numeric",
    minute: "numeric",
    second: "numeric",
    hour12: false,
    timeZoneName: 'short'
  };
  return Intl.DateTimeFormat("en-US", options).format(dateObj);
};
class ProjectIssueStatus extends Component {
  state = {
    issues: null
  };
  componentDidMount() {
    // fetch("https://api.github.com/repos/facebook/react/issues")
    //   .then(response => response.json())
    // .then(issues => this.setState({ issues }));

    this.setState({
      issues: captured_response.map(response => ({
        ...response,
        created_at: formatDate(response.created_at),
        updated_at: formatDate(response.updated_at)
      }))
    });
  }

  render() {
    this.state.issues &&
      this.state.issues.map(({ created_at }) => {
        const createdAtObj = new Date(created_at);
        console.log(createdAtObj.getDate());
      });
    return (
      <Table celled>
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell>Issue Number</Table.HeaderCell>
            <Table.HeaderCell>Title</Table.HeaderCell>
            <Table.HeaderCell>Created At</Table.HeaderCell>
            <Table.HeaderCell>Updated At</Table.HeaderCell>
            <Table.HeaderCell>Labels</Table.HeaderCell>
            <Table.HeaderCell>State</Table.HeaderCell>
          </Table.Row>
        </Table.Header>
        <Table.Body>
          {this.state.issues &&
            this.state.issues.map(issue => {
              const {
                number: IssueNumber,
                title,
                created_at,
                updated_at,
                labels,
                state
              } = issue;
              return (
                <Table.Row key={IssueNumber}>
                  <Table.Cell>{IssueNumber}</Table.Cell>
                  <Table.Cell>{title}</Table.Cell>
                  <Table.Cell>{created_at}</Table.Cell>
                  <Table.Cell>{updated_at}</Table.Cell>
                  <Table.Cell>
                    <Labels labels={labels} />
                  </Table.Cell>
                  <Table.Cell>{state}</Table.Cell>
                </Table.Row>
              );
            })}
        </Table.Body>
      </Table>
    );
  }
}

export default ProjectIssueStatus;
