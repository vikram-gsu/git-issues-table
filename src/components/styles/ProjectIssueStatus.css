.table {
  width: 100%;
  margin: 10px auto;
  border-radius:10px;
  box-shadow: 0 0 2px 1px rgba(0, 0, 0, 0.2);
}

ul {
  list-style-type: none;
  text-align: left;
  margin: 0;
  padding: 0;
}

.pagination {
  display: flex;
  justify-content: center;
}