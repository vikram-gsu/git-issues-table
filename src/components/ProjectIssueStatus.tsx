import React, { Component } from "react";
import { Table, Grid, Search, SearchProps } from "semantic-ui-react";
import "./styles/ProjectIssueStatus.css";
import { GitlabIssue, label } from "./LandingPage";

const escape = (s: string) => {
  return s.replace(/[-\\^$*+?.()|[\]{}]/g, "\\$&");
};

type LabelsProps = {
  labels: [label?];
};
const Labels = ({ labels }: LabelsProps) => (
  <ul>
    {labels && labels.length>0 && labels.map(label => {
      const { url, color, name, id } = label!;
      return (
        <li key={id}>
          <a
            href={url}
            style={{
              backgroundColor: `#${color}`,
              color: "black",
              padding: "0.1em",
              borderRadius: "1px"
            }}
          >
            {name}
          </a>
        </li>
      );
    })}
  </ul>
);

type ProjectIssueStatusProps = {
  issues: Array<GitlabIssue>;
};

type ProjectIssueStatusStatus = {
  issues: Array<GitlabIssue>;
  isLoading: Boolean;
  value?: string;
};

class ProjectIssueStatus extends Component<
  ProjectIssueStatusProps,
  ProjectIssueStatusStatus
> {
  initialState = {
    issues: this.props.issues,
    isLoading: false,
    value: ""
  };
  state = this.initialState;

  handleSearchChange = (
    e: React.MouseEvent<HTMLElement, MouseEvent>,
    { value }: SearchProps
  ) => {
    this.setState({ isLoading: true, value });

    setTimeout(() => {
      if (this.state.value.length < 1) return this.setState(this.initialState);

      const re = new RegExp(escape(this.state.value), "i");
      const isMatch = (issue: GitlabIssue) => re.test(issue.title);

      this.setState({
        isLoading: false,
        issues: this.props.issues.filter(issue => isMatch(issue))
      });
    }, 300);
  };

  render() {
    return (
      <div>
        <Grid columns={16}>
          <Grid.Row>
            <Grid.Column width={2}></Grid.Column>
            <Grid.Column width={4}>
              <h1>Github Issue Details</h1>
            </Grid.Column>
            <Grid.Column width={4}></Grid.Column>
            <Grid.Column width={4}>
              <div className="search">
                <Search
                  size="huge"
                  width="100%"
                  loading={this.state.isLoading}
                  onSearchChange={this.handleSearchChange}
                />
              </div>
            </Grid.Column>
            <Grid.Column width={2}></Grid.Column>
          </Grid.Row>
          <Grid.Column width={2}></Grid.Column>
          <Grid.Column width={12}>
            <div className="table">
              <Table celled>
                <Table.Header>
                  <Table.Row>
                    <Table.HeaderCell>Issue Number</Table.HeaderCell>
                    <Table.HeaderCell>Title</Table.HeaderCell>
                    <Table.HeaderCell>Created At</Table.HeaderCell>
                    <Table.HeaderCell>Updated At</Table.HeaderCell>
                    <Table.HeaderCell>Labels</Table.HeaderCell>
                    <Table.HeaderCell>State</Table.HeaderCell>
                  </Table.Row>
                </Table.Header>
                <Table.Body>
                  {this.state.issues &&
                    this.state.issues.map(issue => {
                      const {
                        number: IssueNumber,
                        title,
                        created_at,
                        updated_at,
                        labels,
                        state
                      } = issue;
                      return (
                        <Table.Row key={IssueNumber}>
                          <Table.Cell>{IssueNumber}</Table.Cell>
                          <Table.Cell>{title}</Table.Cell>
                          <Table.Cell>{created_at}</Table.Cell>
                          <Table.Cell>{updated_at}</Table.Cell>
                          <Table.Cell>
                            <Labels labels={labels} />
                          </Table.Cell>
                          <Table.Cell>{state}</Table.Cell>
                        </Table.Row>
                      );
                    })}
                </Table.Body>
              </Table>
            </div>
          </Grid.Column>
          <Grid.Column width={2}></Grid.Column>
        </Grid>
      </div>
    );
  }
}

export default ProjectIssueStatus;
