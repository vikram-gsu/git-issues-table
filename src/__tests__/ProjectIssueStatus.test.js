import { mount } from "enzyme";
import React from 'react';
import toJSON from 'enzyme-to-json';
import ProjectIssueStatus from "../components/ProjectIssueStatus";

const mockResponses = [
  {
    number: 16630,
    title: "Chrome Extension Background Page and useEffect Not Triggered",
    labels: [
      {
        "id": 196858374,
        "node_id": "MDU6TGFiZWwxOTY4NTgzNzQ=",
        "url": "https://api.github.com/repos/facebook/react/labels/CLA%20Signed",
        "name": "CLA Signed",
        "color": "e7e7e7",
        "default": false
      },
      {
        "id": 1296748300,
        "node_id": "MDU6TGFiZWwxMjk2NzQ4MzAw",
        "url": "https://api.github.com/repos/facebook/react/labels/React%20Flare",
        "name": "React Flare",
        "color": "63f2ff",
        "default": false
      }
    ],
    state: "open",
    created_at: "August 31, 2019, 10:31:51 EDT",
    updated_at: "August 31, 2019, 10:39:19 EDT"
  },
  {
    number: 16629,
    title: "Can't use instance values without refs",
    labels: [],
    state: "open",
    created_at: "August 31, 2019, 10:31:51 EDT",
    updated_at: "August 31, 2019, 10:39:19 EDT"
  }

];

describe("<ProjectIssueStatus />", () => {

  it("renders page without crashing", () => {
    mount(<ProjectIssueStatus issues={mockResponses} />);
  })

  it("renders tableheader as expected", () => {
    const wrapper = mount(<ProjectIssueStatus issues={mockResponses} />);
    console.log(wrapper.debug());

    const TableHeader = wrapper.find('TableHeader');
    expect(TableHeader.children().text()).toContain('Issue Number');
    expect(TableHeader.children().text()).toContain('Title');
    expect(TableHeader.children().text()).toContain('Created At');
    expect(TableHeader.children().text()).toContain('Updated At');
    expect(TableHeader.children().text()).toContain('Labels');
    expect(TableHeader.children().text()).toContain('State');
  })

  it("renders tablebody as expected", () => {
    const mockResponse = mockResponses[0];
    const wrapper = mount(<ProjectIssueStatus issues={mockResponses} />);
    console.log(wrapper.debug());

    const TableBody = wrapper.find('TableBody');
    expect(TableBody.children().text()).toContain(mockResponse.title);

    expect(TableBody.find('TableRow').children()).toHaveLength(2);
    expect(toJSON(TableBody)).toMatchSnapshot();
  });
  
  it("renders labels component as expected", () => {
    let mockResponseWithLabels = mockResponses.filter(response => response.labels.length >0);
    mockResponseWithLabels = (mockResponseWithLabels.length>0)?mockResponseWithLabels[0]: mockResponseWithLabels;
    
    let mockResponseWithoutLabels = mockResponses.filter(response => response.labels.length >0);
    mockResponseWithoutLabels = (mockResponseWithoutLabels.length>0)?mockResponseWithoutLabels[0]: mockResponseWithoutLabels;
    
    const wrapper = mount(<ProjectIssueStatus issues={mockResponses} />);
    console.log(wrapper.debug());

    const Labels = wrapper.find('Labels');
    expect(Labels.children().text()).toContain('React Flare');
    expect(Labels.children()).toHaveLength(2);
    
    expect(toJSON(Labels)).toMatchSnapshot()

  })
});
