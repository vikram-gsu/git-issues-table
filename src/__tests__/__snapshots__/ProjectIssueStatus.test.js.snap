// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectIssueStatus /> renders tablebody as expected 1`] = `
<TableBody
  as="tbody"
>
  <tbody
    className=""
  >
    <TableRow
      as="tr"
      cellAs="td"
      key="16630"
    >
      <tr
        className=""
      >
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            16630
          </td>
        </TableCell>
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            Chrome Extension Background Page and useEffect Not Triggered
          </td>
        </TableCell>
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            August 31, 2019, 10:31:51 EDT
          </td>
        </TableCell>
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            August 31, 2019, 10:39:19 EDT
          </td>
        </TableCell>
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            <Labels
              labels={
                Array [
                  Object {
                    "color": "e7e7e7",
                    "default": false,
                    "id": 196858374,
                    "name": "CLA Signed",
                    "node_id": "MDU6TGFiZWwxOTY4NTgzNzQ=",
                    "url": "https://api.github.com/repos/facebook/react/labels/CLA%20Signed",
                  },
                  Object {
                    "color": "63f2ff",
                    "default": false,
                    "id": 1296748300,
                    "name": "React Flare",
                    "node_id": "MDU6TGFiZWwxMjk2NzQ4MzAw",
                    "url": "https://api.github.com/repos/facebook/react/labels/React%20Flare",
                  },
                ]
              }
            >
              <ul>
                <li
                  key="196858374"
                >
                  <a
                    href="https://api.github.com/repos/facebook/react/labels/CLA%20Signed"
                    style={
                      Object {
                        "backgroundColor": "#e7e7e7",
                        "borderRadius": "1px",
                        "color": "black",
                        "padding": "0.1em",
                      }
                    }
                  >
                    CLA Signed
                  </a>
                </li>
                <li
                  key="1296748300"
                >
                  <a
                    href="https://api.github.com/repos/facebook/react/labels/React%20Flare"
                    style={
                      Object {
                        "backgroundColor": "#63f2ff",
                        "borderRadius": "1px",
                        "color": "black",
                        "padding": "0.1em",
                      }
                    }
                  >
                    React Flare
                  </a>
                </li>
              </ul>
            </Labels>
          </td>
        </TableCell>
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            open
          </td>
        </TableCell>
      </tr>
    </TableRow>
    <TableRow
      as="tr"
      cellAs="td"
      key="16629"
    >
      <tr
        className=""
      >
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            16629
          </td>
        </TableCell>
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            Can't use instance values without refs
          </td>
        </TableCell>
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            August 31, 2019, 10:31:51 EDT
          </td>
        </TableCell>
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            August 31, 2019, 10:39:19 EDT
          </td>
        </TableCell>
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            <Labels
              labels={Array []}
            >
              <ul />
            </Labels>
          </td>
        </TableCell>
        <TableCell
          as="td"
        >
          <td
            className=""
          >
            open
          </td>
        </TableCell>
      </tr>
    </TableRow>
  </tbody>
</TableBody>
`;
